pkg/controller/garbagecollector/rate_limiter_helper.go

```diff
@@ -29,31 +29,32 @@ import (
 // RegisteredRateLimiter records the registered RateLimters to avoid
 // duplication.
 type RegisteredRateLimiter struct {
-	rateLimiters map[unversioned.GroupVersion]struct{}
-	lock         sync.RWMutex
+	rateLimiters map[unversioned.GroupVersion]*sync.Once
 }
 
 // NewRegisteredRateLimiter returns a new RegisteredRateLimiater.
-func NewRegisteredRateLimiter() *RegisteredRateLimiter {
-	return &RegisteredRateLimiter{
-		rateLimiters: make(map[unversioned.GroupVersion]struct{}),
+// TODO: NewRegisteredRateLimiter is not dynamic. We need to find a better way
+// when GC dynamically change the resources it monitors.
+func NewRegisteredRateLimiter(resources []unversioned.GroupVersionResource) *RegisteredRateLimiter {
+	rateLimiters := make(map[unversioned.GroupVersion]*sync.Once)
+	for _, resource := range resources {
+		gv := resource.GroupVersion()
+		if _, found := rateLimiters[gv]; !found {
+			rateLimiters[gv] = &sync.Once{}
+		}
 	}
+	return &RegisteredRateLimiter{rateLimiters: rateLimiters}
 }
 
 func (r *RegisteredRateLimiter) registerIfNotPresent(gv unversioned.GroupVersion, client *dynamic.Client, prefix string) {
-	r.lock.RLock()
```

Comment:
I don't buy it. We're calling this a max of 100 times per second (due to
rate limiter). Is RLock really that bad??

On Mon, Aug 22, 2016 at 11:25 PM, Wojciech Tyczynski <
notifications@github.com> wrote:

> In pkg/controller/garbagecollector/rate_limiter_helper.go
> <https://github.com/kubernetes/kubernetes/pull/31044#discussion_r75808550>
> :
>
> >  }
> >
> >  func (r *RegisteredRateLimiter) registerIfNotPresent(gv unversioned.GroupVersion, client *dynamic.Client, prefix string) {
> > -	r.lock.RLock()
>
> The reason why it affected performance is that this method
> (registerIfNotPresent) is called from:
> GC::deleteObject, GC::getObject and GC::updateObject.
>
> Even if RLock is cheap, it consumes some cpu, and if we are calling it all
> the time, it will sum up to something visible.
>
> —
> You are receiving this because you were assigned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/31044/files/992afd9c45f6f266cb753f96ddae4b54786dd789#r75808550>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAnglh3t9-u_XaIVcL91820kLpVQAkQ3ks5qipJvgaJpZM4Jo8eq>
> .
>